### PR TITLE
Update element_type inference (default_type_hints) for batched DoFns with yields_batches/yields_elements  

### LINKED ISSUE #22197: [Bug]: DoFn.process with @yields_batches produces PCollection with incorrect element_type
### What happened?

When determining the element_type for the PCollection output from a DoFn.process with `@yields_batchhes`, we incorrectly use the _batch_ typehint. This affects Beam 2.40.0.

Relates to #21656
Part of #21650

For example: when applying a DoFn like:

```py
class ElementToBatchDoFn(beam.DoFn):                                            
   @beam.DoFn.yields_batches                                                     
   def process(self, element: int) -> Iterator[List[int]]:      
     yield [element] * element                                                   
                                                                                 
   def infer_output_type(self, input_element_type):                              
     return input_element_type 
```

We indicated that the output PCollection has `element_type` `List[int]` (the batch type), instead of `int` (the element type).

### Issue Priority

Priority: 2

### Issue Component

Component: sdk-py-core

### PR DESCRIPTION
**Pull request by TheNeuralBit**
Fixes #22197 

This updates `DoFn.default_type_hints` to inspect the output type annotation for the appropriate method (process, or process_batch) depending on the yields_batches/yields_elements decorators.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/transforms/core.py [hunk 1]
```diff
@@ -708,6 +708,36 @@ def get_function_arguments(self, func):
 
   def default_type_hints(self):
     fn_type_hints = typehints.decorators.IOTypeHints.from_callable(self.process)
+    batch_fn_type_hints = typehints.decorators.IOTypeHints.from_callable(
+        self.process_batch)
+
+    if fn_type_hints is not None:
```
Comment (ryanthompson591) [inline sdks/python/apache_beam/transforms/core.py:714]: This nested logic is hard to read and a little messy and hard to follow.

For example here and below we have "if fn_type_hints is not None:"

I had to write this a little to really wrap my head around it. I don't know if any parts of my rewrite are useful.

```
# default type hints can come from one of two places. The process type hints or the process batch type hints.
# also the process might be set up to yield batches or to yield elements

has_fn_type_hints = fn_type_hints is not None
has_batch_type_hints = batch_fn_type_hints is not None
yields_batches = self._process_yields_batches
yeilds_elements = self._process_yields_elements

if has_fn_type_hints:
  if yields_elements and has_batch_type_hints:
    # because both batch and element type hints are defined, raise an exception if they are not equal.
    self.validate_batch_and_fn_typehints_same()
    fn_type_hints = fn_type_hints.with_output_types_from(
            batch_fn_type_hints)  # TODO: comment why this is done.
  elif yields_batches:
    # Because the process method produces batches, don't use its output typehints.
    fn_type_hints = fn_type_hints.with_output_types_from(
            typehints.decorators.IOTypeHints.empty())
elif yields_elements and has_batch_type_hints:
  # process_batch produces elements, grab its output typehint
  fn_type_hints = batch_fn_type_hints.with_input_types_from(
            typehints.decorators.IOTypeHints.empty())  # TODO comment why does this use with_input_types_from
```

**Review by TheNeuralBit — COMMENTED** [review]
Thanks for the feedback!

File: sdks/python/apache_beam/transforms/core.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/transforms/core.py]: I was a little concerned rewriting this way could break some esoteric use-case, but I ran all our internal tests and it looks good.

**Comment by TheNeuralBit** [comment]
assign to next reviewer

**Comment by TheNeuralBit** [comment]
R: @tvalentyn could you take a look? It looks like Ryan will be out through next week and I'd like to get this in for the release cut.

**Review by tvalentyn — APPROVED** [review]

**Comment by tvalentyn** [comment]
I checked that `get_type_hints(object).with_defaults(beam.typehints.decorators.IOTypeHints.empty().strip_iterable())` is the same as `get_type_hints(object).with_defaults(None)`. Looks like this doesn't change evaluation for non-batching DoFn's, so shouldn't be a breaking change. But you can also try to import and run TGP if you think this needs more coverage.

**Comment by TheNeuralBit** [comment]
Thanks for checking on that! I ran a TGP and everything looks good.
